winpr/libwinpr/utils/wlog/UdpAppender.c

```diff
@@ -0,0 +1,209 @@
+		return FALSE;
+	}
+
+	memcpy(&appender->targetAddr, result->ai_addr, result->ai_addrlen);
+	appender->targetAddrLen = result->ai_addrlen;
+
+
+	return TRUE;
+}
+
+BOOL Wlog_UdpAppender_SetTarget(wLogUdpAppender* appender, const char *host)
+{
+
+	appender->targetAddrLen = 0;
+	if (appender->host)
+		free(appender->host);
+
+	appender->host = _strdup(host);
+	return (appender->host != NULL) && WLog_UdpAppender_Open(NULL, appender);
+}
+
+static BOOL WLog_UdpAppender_Close(wLog* log, wLogUdpAppender* appender)
+{
+	if (!log || !appender)
+		return FALSE;
+
+	return TRUE;
+}
+
+static BOOL WLog_UdpAppender_WriteMessage(wLog* log, wLogUdpAppender* appender, wLogMessage* message)
+{
+	char prefix[WLOG_MAX_PREFIX_SIZE];
+
+	if (!log || !appender || !message)
+		return FALSE;
+
+	message->PrefixString = prefix;
+	WLog_Layout_GetMessagePrefix(log, appender->Layout, message);
+
+	_sendto(appender->sock, message->PrefixString, strlen(message->PrefixString),
+				0, &appender->targetAddr, appender->targetAddrLen);
+
+	_sendto(appender->sock, message->TextString, strlen(message->TextString),
+						0, &appender->targetAddr, appender->targetAddrLen);
+
+	_sendto(appender->sock, "\n", 1, 0, &appender->targetAddr, appender->targetAddrLen);
+
+	return TRUE;
+}
+
+static BOOL WLog_UdpAppender_WriteDataMessage(wLog* log, wLogUdpAppender* appender, wLogMessage* message)
+{
+	if (!log || !appender || !message)
+		return FALSE;
+
+	return TRUE;
+}
+
+static BOOL WLog_UdpAppender_WriteImageMessage(wLog* log, wLogUdpAppender* appender, wLogMessage* message)
+{
+	if (!log || !appender || !message)
+		return FALSE;
+
+
+	return TRUE;
+}
+
+wLogUdpAppender* WLog_UdpAppender_New(wLog* log)
+{
+	wLogUdpAppender* appender;
+	DWORD nSize;
+	LPCSTR name;
+
+	appender = (wLogUdpAppender*) calloc(1, sizeof(wLogUdpAppender));
+	if (!appender)
+		return NULL;
+
+	appender->Type = WLOG_APPENDER_UDP;
+
+	appender->Open = (WLOG_APPENDER_OPEN_FN) WLog_UdpAppender_Open;
+	appender->Close = (WLOG_APPENDER_OPEN_FN) WLog_UdpAppender_Close;
+	appender->WriteMessage =
+			(WLOG_APPENDER_WRITE_MESSAGE_FN) WLog_UdpAppender_WriteMessage;
+	appender->WriteDataMessage =
+			(WLOG_APPENDER_WRITE_DATA_MESSAGE_FN) WLog_UdpAppender_WriteDataMessage;
+	appender->WriteImageMessage =
+			(WLOG_APPENDER_WRITE_IMAGE_MESSAGE_FN) WLog_UdpAppender_WriteImageMessage;
+
+	appender->sock = _socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+	if (appender->sock == INVALID_SOCKET)
+		goto error_sock;
+
+	name = "WLOG_UDP_TARGET";
+	nSize = GetEnvironmentVariableA(name, NULL, 0);
+	if (nSize)
+	{
+		appender->host = (LPSTR) malloc(nSize);
+		if (!appender->host)
+			goto error_env_malloc;
+
+		GetEnvironmentVariableA(name, appender->host, nSize);
+
+		if (!WLog_UdpAppender_Open(log, appender))
+			goto error_open;
+	}
+	else
+	{
+		appender->host = _strdup("127.0.0.1:20000");
+		if (!appender->host)
+			goto error_env_malloc;
+	}
+
+	return appender;
+
+error_open:
+	free(appender->host);
+error_env_malloc:
```

Comment:
unimportant remark: rename label to error_host_alloc since it is also targeted from strdup